File: 2719-count-of-integers/2719-count-of-integers.py
mod = 10**9 + 7

def digitDP(s, minSum, maxSum):
	n = len(s)

	dp = [[[0]*(maxSum+20) for _ in range(2)] for _ in range(n+1)]

	# Empty prefix with sum=0
	dp[0][1][0] = 1

	for i in range(n):
		for tight in range(2):
			for sm in range(maxSum+1):
				up = int(s[i]) if tight else 9
				for digit in range(up + 1):
					dp[i+1][tight and (digit == up)][sm+digit] += dp[i][tight][sm]
					dp[i+1][tight and (digit == up)][sm+digit] %= mod

	res = 0
	for i in range(minSum, maxSum+1):
		res += (dp[n][1][i] + dp[n][0][i])
		res %= mod

	return res % mod


class Solution:
	def count(self, num1: str, num2: str, minSum: int, maxSum: int) -> int:
		res = digitDP(num2,minSum,maxSum) 
		res -= digitDP(str(int(num1)-1),minSum,maxSum)
		return res % mod

File: 2719-count-of-integers/2719-count-of-integers.cpp
class Solution {
public:
    
    int MOD = 1e9+7;
    int memo[23][2][200];
        
    void subtract(string &num){
        
        int N = num.size();
        
        for(int i=N-1; i>=0; i--){
                
            if(num[i]=='0')
                num[i]='9';
            else{
                num[i]--;
                break;
            }
        }
    }
    
    int solve(string num, int idx, bool last, int dig_sum, int max_sum, int min_sum){
        
        if(idx==num.size()){
         
            if(dig_sum<=max_sum && dig_sum>=min_sum){
                return 1;
            }else{
                return 0;
            }
        }
        
        
        if(memo[idx][last][dig_sum]!=-1)
            return memo[idx][last][dig_sum];
        
        long long int ans =0;
        int limit = last ? num[idx]-'0':9;
        for(int di=0; di<=limit; di++){
            
            
            ans =( ans + solve(num, idx+1, (last && di==limit), dig_sum+di, max_sum, min_sum) )%MOD;
            
        }
        
        return memo[idx][last][dig_sum] = ans;
    }
    
    int count(string num1, string num2, int min_sum, int max_sum) {
        
        subtract(num1);
        //cout<<num2<<" "<<num1<<endl;
        memset(memo, -1, sizeof(memo));
        int ans2 = solve(num2, 0, true, 0, max_sum, min_sum);
        
        memset(memo, -1, sizeof(memo));
        int ans1 = solve(num1, 0, true, 0, max_sum, min_sum);
        
        //cout<<ans2<<" - "<<ans1<<endl;
        return (ans2-ans1+MOD)%MOD;
    }
};